migration/ubuntu/20-to-24.md

```diff
@@ -0,0 +1,250 @@
+>
+> - If it fails early for some and succeed for others then remove the removed nodes from inventory and retry.
+> - If it fails late for some with timeouts then ignore.
+
+- Check that old etcd members are removed:
+
+  ```bash
+  HOSTNAME="$(kubectl get node -l node-role.kubernetes.io/control-plane= -ojsonpath='{.items[0].metadata.name}')"
+  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
+    etcdctl --endpoints=https://127.0.0.1:2379 \
+            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
+            --cert=/etc/kubernetes/ssl/etcd/server.crt \
+            --key=/etc/kubernetes/ssl/etcd/server.key \
+            member list
+  ```
+
+- If there are old nodes left, remove them using the `uuid` from the previous command (The id in the first column of each row)
+
+  ```bash
+  HOSTNAME="$(kubectl get node -l node-role.kubernetes.io/control-plane= -ojsonpath='{.items[0].metadata.name}')"
+  kubectl -n kube-system exec -it "etcd-${HOSTNAME}" -- \
+    etcdctl --endpoints=https://127.0.0.1:2379 \
+            --cacert=/etc/kubernetes/ssl/etcd/ca.crt \
+            --cert=/etc/kubernetes/ssl/etcd/server.crt \
+            --key=/etc/kubernetes/ssl/etcd/server.key \
+            member remove <uuid>
+  ```
+
+- Check that old control plane nodes are removed:
+
+  ```bash
+  kubectl get nodes
+  ```
+
+- If there are any old nodes left, delete them
+
+  ```bash
+  kubectl delete nodes <node>,...
+  ```
+
+- Change back to using dynamic inventory
+
+  ```bash
+  pushd "${CK8S_CONFIG_PATH}"
+  cp <sc|wc>-config-backup/inventory.ini <sc|wc>-config/inventory.ini
+  popd
+  ```
+
+- Remove old control plane nodes using Terraform
+- Remove the old control plane floating IPs from group vars under the key `supplementary_addresses_in_ssl_keys`
+
+> [!NOTE]
+> This is not needed if you are using a loadbalancer in front of the API server.
+
+- Apply kubespray:
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray apply <sc|wc> --tags=master,nginx -b
+  popd
+  ```
+
+- Update the NetworkPolicies:
+
+  ```bash
+  pushd "${CK8S_APPS_REPOSITORY_PATH}"
+  ck8s update-ips <sc|wc> apply
+
+  ck8s ops helmfile <sc|wc> -l app=netpol -i apply
+  popd
+  ```
+
+### Replace worker nodes
+
+- Add a new worker node that doesn't specify `image_id` (To use the new image)
+- Verify by running `terraform plan` and make sure there's a new node added with the correct image
+- Add the rest of the new worker nodes
+- Run `terraform apply` to add all the new nodes
+- Add all the new worker nodes
+
+  ```bash
+  pushd "${CK8S_KUBESPRAY_REPOSITORY_PATH}"
+  ck8s-kubespray run-playbook <sc|wc> facts.yml
+  ck8s-kubespray run-playbook <sc|wc> scale.yml -b --limit=<new-node1>,<new-node2>,...
+  popd
+  ```
+
+- Drain the old nodes and make sure everything starts up properly
+
+> [!WARNING]
+> If you are using applications that allowlist the old nodes, please update that list to also include the new nodes before draining the old ones
```

Comment:
```suggestion
> [!caution]
> If you are using applications that allowlist the old nodes, please update that list to also include the new nodes before draining the old ones.
```

Make this harder because here comes the impact :slightly_smiling_face:

Reply:
https://github.com/elastisys/compliantkubernetes-kubespray/pull/409/commits/ac762a4ee1438cd18875d2979e8fcbff1b1cf23e